lighthouse-core/gather/gatherers/seo/font-size.js

```diff
@@ -304,6 +239,94 @@ class FontSize extends Gatherer {
     return {analyzedFailingNodesData, analyzedFailingTextLength};
   }
 
+  /**
+   * @param {LH.Crdp.DOMSnapshot.CaptureSnapshotResponse} snapshot
+   * @return {BackendIdsToFontData}
+   */
+  calculateBackendIdsToFontData(snapshot) {
+    const strings = snapshot.strings;
+
+    // The document under analysis is the root document.
+    const doc = snapshot.documents[0];
+
+    // doc is a flattened property list describing all the Nodes in a document, with all string values
+    // deduped in a strings array.
+    // Implementation:
+    // https://cs.chromium.org/chromium/src/third_party/blink/renderer/core/inspector/inspector_dom_snapshot_agent.cc?sq=package:chromium&g=0&l=534
+
+    // satisfy the type checker that all expected values exist
+    if (!doc || !doc.nodes.nodeType || !doc.nodes.nodeName || !doc.nodes.backendNodeId
+      || !doc.nodes.nodeValue || !doc.nodes.parentIndex) {
+      throw new Error('Unexpected response from DOMSnapshot.captureSnapshot.');
+    }
+
+    // Not all nodes have computed styles (ex: TextNodes), so doc.layout.* is smaller than doc.nodes.*
+    // doc.layout.nodeIndex maps the index into doc.layout.styles to an index into doc.nodes.*
+    // nodeIndexToStyleIndex inverses that mapping.
+    /** @type {Map<number, number>} */
+    const nodeIndexToStyleIndex = new Map();
+    for (let i = 0; i < doc.layout.nodeIndex.length; i++) {
+      nodeIndexToStyleIndex.set(doc.layout.nodeIndex[i], i);
+    }
+
+    /** @type {BackendIdsToFontData} */
+    const backendIdsToFontData = new Map();
+    for (let i = 0; i < doc.nodes.nodeType.length; i++) {
+      const nodeType = doc.nodes.nodeType[i];
+      const nodeValue = strings[doc.nodes.nodeValue[i]];
+      if (!isTextNode({
+        nodeType,
+        parentNodeName: strings[doc.nodes.nodeName[doc.nodes.parentIndex[i]]],
```

Comment:
still would be good to pull this into a variable (or variables), especially to make the conditional obvious at a glance

Reply:
to be clear, do you mean
```
const isTextNode = !isTextNode({
        nodeType,
        parentNodeName: strings[doc.nodes.nodeName[doc.nodes.parentIndex[i]]],
      })
if (!isTextNode) {
  continue;
}
```

or something else?

Reply:
I assumed he meant breaking up `strings[doc.nodes.nodeName[doc.nodes.parentIndex[i]]]` but even if he didn't, I think that would be nice if it isn't too much trouble :)